### PR TITLE
Elasticsearch end-to-end regression tests

### PR DESCRIPTION
**Pull request by lintool**
hey @r-clancy can you give this a whirl? Should be pretty straightforward. Subcommands:

```bash
python src/main/python/run_es_regression.py --ping
python src/main/python/run_es_regression.py --check-index-exists robust04
python src/main/python/run_es_regression.py --delete-index robust04
python src/main/python/run_es_regression.py --create-index robust04
python src/main/python/run_es_regression.py --insert-docs robust04 --input /tuna1/collections/newswire/disk45/
python src/main/python/run_es_regression.py --evaluate robust04
```

This strings everything together:

```bash
python src/main/python/run_es_regression.py --regression robust04 --input /tuna1/collections/newswire/disk45/
```
And for MS MARCO passage:

```bash
python src/main/python/run_es_regression.py --ping
python src/main/python/run_es_regression.py --check-index-exists msmarco-passage
python src/main/python/run_es_regression.py --delete-index msmarco-passage
python src/main/python/run_es_regression.py --create-index msmarco-passage
python src/main/python/run_es_regression.py --insert-docs msmarco-passage --input /tuna1/collections/msmarco/passage
python src/main/python/run_es_regression.py --evaluate msmarco-passage

# Run end to end
python src/main/python/run_es_regression.py --regression msmarco-passage --input /tuna1/collections/msmarco/passage
```

Once you sign off, let me merge so I can write a detailed commit message?

### CODE REVIEW

File: src/main/python/run_es_regression.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+    def create_index(self, collection):
+        logger.info('Creating index {}...'.format(collection))
+        # Make sure the index does not exist:
+        if not self.does_index_exist(collection):
+            filename = 'src/main/resources/elasticsearch/index-config.{}.json'.format(collection)
+            if not os.path.exists(filename):
+                raise Exception('No config found in src/main/resources/elasticsearch/ for {}!'.format(collection))
+            logger.info('Using index config for {} at {}'.format(collection, filename))
+            file = open(filename, mode='r')
```
Comment (r-clancy) [inline src/main/python/run_es_regression.py]: Python also has the `with` construct that automatically closes the file, it's the more Python-ic way...:
```
with open(filename, mode='r') as file:
    json = file.read()
```

Comment (lintool) [reply src/main/python/run_es_regression.py]: ah, good call!

**Comment by lintool** [comment]
@r-clancy thanks. I'll wait for @x389liu to replicate also.

**Comment by x389liu** [comment]
> @r-clancy thanks. I'll wait for @x389liu to replicate also.

sure, I'm doing this today :)

File: src/main/python/run_es_regression.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        if math.isclose(ap, expected): logger.info('[SUCESS] {} MAP verified as expected!'.format(ap))
+        else: logger.info('[FAILED] {} MAP, expected {} MAP!'.format(ap, expected))
+
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Program for running Elasticsearch regressions.')
+    parser.add_argument('--ping', action='store_true', default=False, help='ping ES and exit')
+    parser.add_argument('--check-index-exists', default='', type=str, metavar='collection', help='check if index exists')
+    parser.add_argument('--delete-index', default='', type=str, metavar='collection', help='check if index exists')
```
Comment (x389liu) [inline src/main/python/run_es_regression.py]: forgot to change help information here?

Comment (lintool) [reply src/main/python/run_es_regression.py]: good catch, fixed.

File: src/main/python/run_es_regression.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            raise Exception('Location of corpus not specified (use --input)!')
+        if not es.is_alive():
+            raise Exception('Elasticsearch does not appear to be alive!')
+        if es.does_index_exist(args.regression):
+            logger.info('Index {} already exists: deleting and recreating.'.format(args.regression))
+            es.delete_index(args.regression)
+        es.create_index(args.regression)
+        es.insert_docs(args.regression, args.input)
+        logger.info('Document ingestion complete. Sleeping now for 60s...')
+        time.sleep(60)
```
Comment (x389liu) [inline src/main/python/run_es_regression.py]: question.. why do we want to pause here?

Comment (lintool) [reply src/main/python/run_es_regression.py]: This is to wait for ES to "catch up" - ingested documents are not available to search immediately. @r-clancy tried tweaking various refresh periods, etc., but the safest way is to just wait. I'll add a comment to document.

Comment (x389liu) [reply src/main/python/run_es_regression.py]: I see, thanks for the explanation!

File: src/main/python/run_es_regression.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            except:
+                logger.info(response)
+                return False
+            else:
+                return True
+        else:
+            raise Exception('The index {} already exists!'.format(collection))
+
+    def insert_docs(self, collection, path):
+        logger.info('Inserting documents from {} into {}... '.format(path, collection))
```
Comment (x389liu) [inline src/main/python/run_es_regression.py]: One more small thing: do we want to check if `path` here is valid? And throw an exception if not.
For example, when I tried running `python src/main/python/run_es_regression.py --regression msmarco-passage --input random_non_existed_path`, the program throw an exception from java, but the python program resumes: sleeping for 60s, do evaluation, and finally give me a runtime error..

Comment (lintool) [reply src/main/python/run_es_regression.py]: Good catch. Will add.

**Pull request merged**